File: go/0015.3sum/solution.go
// Created by Olgierd Palasz at 2026/06/22 20:38
// leetgo: dev
// https://leetcode.com/problems/3sum/

package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func threeSum(nums []int) [][]int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})

	results := [][]int{}

	for i, val := range nums {
		if i > 0 && nums[i-1] == nums[i] {
			continue
		}

		if val > 0 {
			break
		}

		l := i + 1
		r := len(nums) - 1

		for l < r {
			if val+nums[l]+nums[r] < 0 {
				l += 1
			} else if val+nums[l]+nums[r] > 0 {
				r -= 1
			} else {
				results = append(results, []int{val, nums[l], nums[r]})
				l += 1
				r -= 1
				for l < r && nums[l] == nums[l-1] {
					l += 1
				}
				for l < r && nums[r] == nums[r+1] {
					r -= 1
				}
			}
		}
	}

	return results
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	ans := threeSum(nums)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0015.3sum/solution.rs
// Created by Olgierd Palasz at 2026/07/23 19:18
// leetgo: dev
// https://leetcode.com/problems/3sum/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::{collections::HashSet, println};
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort_unstable();

        let mut left: usize;
        let mut right: usize;

        let mut result: HashSet<[i32; 3]> = HashSet::new();

        for (i, number) in nums.iter().enumerate() {
            left = i + 1;
            right = nums.len() - 1;

            while left < right {
                if nums[left] + nums[right] + number < 0 {
                    left += 1;
                } else if nums[left] + nums[right] + number > 0 {
                    right -= 1;
                } else {
                    result.insert([nums[left], *number, nums[right]]);
                    left += 1;
                }
            }
        }

        result.into_iter().map(|t| t.to_vec()).collect()
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let ans: Vec<Vec<i32>> = Solution::three_sum(nums);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
